windows/scripts/extract_kernels.py

```diff
@@ -0,0 +1,201 @@
+#!/usr/bin/env python
+
+# Copyright (C) 2017, Adrian Herrera
+# All rights reserved.
+#
+# Licensed under the MIT License.
+
+"""
+Extracts the Windows kernel executable (``ntoskrnl.exe``) from an ISO. This is
+useful when updating the ``winmonitor_gen.c`` file, as the kernel executable
+is used to generate the correct data structure offsets for that particular
+kernel.
+
+Usage
+-----
+
+::
+
+    python extract_kernels.py -d /path/to/isos -o /path/to/output
+
+Where ``isos`` is a directory containing all or some of the supported ISOs
+listed in s2e/guest-images/images.json and ``output`` is the directory to store
+the kernel executable files.
+
+Requirements
+------------
+
+Requires 7-Zip (http://www.7-zip.org/) and that the ``7z`` executable is
+available on the ``PATH``.
+
+This script will work with both Python 2.7 and 3.x.
+"""
+
+
+from __future__ import print_function
+
+
+import argparse
+import os
+import shutil
+import subprocess
+
+try:
+    from subprocess import DEVNULL
+except ImportError:
+    DEVNULL = open(os.devnull, 'wb')
+
+import sys
+import tempfile
+
+
+# Maps Windows ISOs (taken from the supported images listed in
+# s2e/guest-images/images.json) to the location of ntoskrnl.exe within the ISO.
+#
+# ntoskrnl.exe is stored within an intermediary container. The path to this
+# intermediary container is given in the first element in the value tuple. The
+# second tuple element is the path to ntoskrnl.exe within this intermediary
+# container.
+NTOSKRNL_MAP = {
+    'en_windows_xp_professional_with_service_pack_3_x86_cd_x14-80428.iso':
+        (os.path.join('I386', 'NTOSKRNL.EX_'),
+         'ntoskrnl.exe'),
+    'en_windows_7_enterprise_with_sp1_x64_dvd_u_677651.iso':
+        (os.path.join('sources', 'install.wim'),
+         os.path.join('Windows', 'System32', 'ntoskrnl.exe')),
```

Comment:
They are not guaranteed to be in system32. You must search across the entire archive.

Reply:
Really? In all of these ISOs they were all found in System32.

Reply:
Unless you also mean I have to handle SysWow64? The README says nothing about these, so what would I (as a user) do with the SysWow64 kernel executables? Do I also run symchk + pdbparser on it and add the results to winmonitor_gen.c?

Reply:
AFAIK, the kernel in system32 can be customized depending on your system. The specific kernel is somewhere in the system caches in the windows folder.

Reply:
You may also need to check for *.ex_ in all ISOs, caches might be compressed.
Instead of hard-coding locations for every OS, just look for the kernel file name patterns everywhere in the ISO.

Reply:
Interesting, I did not know that! I will search the entire archive.

Reply:
Here is an example on my system:
```
c:\windows\system32\ntoskrnl.exe
c:\windows\winsxs\amd64_microsoft-windows-os-kernel_31bf3856ad364e35_10.0.14393.0_none_99a4cd1bcad5b6cf\ntoskrnl.exe
c:\windows\winsxs\amd64_microsoft-windows-os-kernel_31bf3856ad364e35_10.0.14393.1198_none_e1f58a060eafce6a\ntoskrnl.exe
```

Reply:
Ok cool, I didn't even know these things existed!